cpp/tests/utilities_tests/type_check_tests.cpp

```diff
@@ -0,0 +1,219 @@
+
+  LCW lhs3{{LCW{1}, LCW{2, 3}}}, rhs3{{LCW{4, 5}}};
+  EXPECT_TRUE(column_types_equal(lhs3, rhs3));
+
+  LCW lhs4{{LCW{1}, LCW{}, LCW{2, 3}}}, rhs4{{LCW{4, 5}, LCW{}}};
+  EXPECT_TRUE(column_types_equal(lhs4, rhs4));
+}
+
+TYPED_TEST(ColumnTypeCheckTestTyped, SameDictionary)
+{
+  using DCW = dictionary_column_wrapper<TypeParam>;
+  DCW lhs{1, 1, 2, 3}, rhs{5, 5};
+  EXPECT_TRUE(column_types_equal(lhs, rhs));
+
+  DCW lhs2{1, 1, 2, 3}, rhs2{};
+  EXPECT_TRUE(column_types_equal(lhs2, rhs2));
+
+  DCW lhs3{}, rhs3{};
+  EXPECT_TRUE(column_types_equal(lhs3, rhs3));
+}
+
+TEST_F(ColumnTypeCheckTest, SameStruct)
+{
+  using SCW      = structs_column_wrapper;
+  using FCW      = fixed_width_column_wrapper<int32_t>;
+  using StringCW = strings_column_wrapper;
+  using LCW      = lists_column_wrapper<int32_t>;
+  using DCW      = dictionary_column_wrapper<int32_t>;
+
+  FCW lf1{1, 2, 3}, rf1{0, 1};
+  StringCW lf2{"a", "bb", ""}, rf2{"cc", "d"};
+  LCW lf3{LCW{1, 2}, LCW{}, LCW{4}}, rf3{LCW{1}, LCW{2}};
+  DCW lf4{5, 5, 5}, rf4{9, 9};
+
+  SCW lhs{lf1, lf2, lf3, lf4}, rhs{rf1, rf2, rf3, rf4};
+  EXPECT_TRUE(column_types_equal(lhs, rhs));
+}
+
+TEST_F(ColumnTypeCheckTest, DifferentBasics)
+{
+  fixed_width_column_wrapper<int32_t> lhs1{1, 1};
+  strings_column_wrapper rhs1{"a", "bb"};
+
+  EXPECT_FALSE(column_types_equal(lhs1, rhs1));
+
+  lists_column_wrapper<string_view> lhs2{{"hello"}, {"world", "!"}};
+  strings_column_wrapper rhs2{"", "kk"};
+
+  EXPECT_FALSE(column_types_equal(lhs2, rhs2));
+
+  fixed_width_column_wrapper<int32_t> lhs3{1, 1};
+  dictionary_column_wrapper<int32_t> rhs3{2, 2};
+
+  EXPECT_FALSE(column_types_equal(lhs3, rhs3));
+
+  lists_column_wrapper<int32_t> lhs4{{8, 8, 8}, {10, 10}};
+  structs_column_wrapper rhs4{rhs2, rhs3};
+
+  EXPECT_FALSE(column_types_equal(lhs4, rhs4));
+}
+
+TEST_F(ColumnTypeCheckTest, DifferentFixedWidth)
+{
+  fixed_width_column_wrapper<int32_t> lhs1{1, 1};
+  fixed_width_column_wrapper<int64_t> rhs1{2};
+
+  EXPECT_FALSE(column_types_equal(lhs1, rhs1));
+
+  fixed_width_column_wrapper<float> lhs2{1, 1};
+  fixed_width_column_wrapper<double> rhs2{2};
+
+  EXPECT_FALSE(column_types_equal(lhs2, rhs2));
+
+  fixed_width_column_wrapper<timestamp_ms> lhs3{1, 1};
+  fixed_width_column_wrapper<timestamp_us> rhs3{2};
+
+  EXPECT_FALSE(column_types_equal(lhs3, rhs3));
+
+  fixed_width_column_wrapper<duration_D> lhs4{};
+  fixed_width_column_wrapper<duration_us> rhs4{42};
+
+  EXPECT_FALSE(column_types_equal(lhs4, rhs4));
+
+  // Same rep, different scale
+  fixed_point_column_wrapper<int32_t> lhs5({10000}, numeric::scale_type{-3});
+  fixed_point_column_wrapper<int32_t> rhs5({10000}, numeric::scale_type{0});
+
+  EXPECT_FALSE(column_types_equal(lhs5, rhs5));
+
+  // Different rep, same scale
+  fixed_point_column_wrapper<int32_t> lhs6({10000}, numeric::scale_type{-1});
+  fixed_point_column_wrapper<int64_t> rhs6({4200}, numeric::scale_type{-1});
+
+  EXPECT_FALSE(column_types_equal(lhs6, rhs6));
```

Comment:
Perhaps add a test for matching fixed-point types too?

Reply:
For matching case I think they are captured in `fixed_width` type lists:
```
[----------] 2 tests from ColumnTypeCheckTestTyped/21, where TypeParam = numeric::fixed_point<int, (numeric::Radix)10>
[ RUN      ] ColumnTypeCheckTestTyped/21.SameFixedWidth
[       OK ] ColumnTypeCheckTestTyped/21.SameFixedWidth (0 ms)
[ RUN      ] ColumnTypeCheckTestTyped/21.SameDictionary
[       OK ] ColumnTypeCheckTestTyped/21.SameDictionary (4 ms)
[----------] 2 tests from ColumnTypeCheckTestTyped/21 (4 ms total)

[----------] 2 tests from ColumnTypeCheckTestTyped/22, where TypeParam = numeric::fixed_point<long, (numeric::Radix)10>
[ RUN      ] ColumnTypeCheckTestTyped/22.SameFixedWidth
[       OK ] ColumnTypeCheckTestTyped/22.SameFixedWidth (0 ms)
[ RUN      ] ColumnTypeCheckTestTyped/22.SameDictionary
[       OK ] ColumnTypeCheckTestTyped/22.SameDictionary (4 ms)
[----------] 2 tests from ColumnTypeCheckTestTyped/22 (4 ms total)
```
It tests for both rep type with radix 10. Do you mean we should test in addition to this?

Reply:
No, this is good. I forgot the fixed-width types include fixed-point.